### ISSUE TITLE
Secrets signatures (for exclusions) should be computed with line (or byte offset) not just filename

### ISSUE THREAD

**Issue by pmevzek-godaddy** [godaddy/tartufo#180]
## Is your feature request related to a problem? Please describe.

My description of the problem I see that could be converted as a new feature (not sure how much it can be seen as a bug or not, it may be subjective, it is not a bug in the sense that code works as expected, but may be a design bug) may seem convoluted (mostly because I lack imagination to find a more stricking case) but please bear with me at least for that example:

```
$ git init test
$ cd test
$ git commit --allow-empty --allow-empty-message -m 'Start'
$ cat howto.txt
This foobar service requires that all API keys start with this string:
zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST
$ git add howto.txt
$ tartufo pre-commit
~~~~~~~~~~~~~~~~~~~~~
Reason: High Entropy
Filepath: howto.txt
Signature: c0c811fad6dddbaf873b4e4c4c9a24879bceaa52dbe762d111fbd0604d1ff85c
@@ -0,0 +1,2 @@
+This foobar service requires that all API keys start with this string:
+zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST

~~~~~~~~~~~~~~~~~~~~~

# The above is as expected, so I add the signature as exclusion

$ tartufo --exclude-signatures c0c811fad6dddbaf873b4e4c4c9a24879bceaa52dbe762d111fbd0604d1ff85c pre-commit
Time: 2021-03-24T16:37:02.161361
All clear. No secrets detected.

# As expected, everything is ok now thanks to the signature exclusion.
# But now let us say the documentation is improved, and by error it includes a real API key!
$ cat howto.txt
This foobar service requires that all API keys start with this string:
zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST


You can use this key for example:
zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST_xR6uDIpycfCfk3yUx

$ git add howto.txt
$ tartufo --exclude-signatures c0c811fad6dddbaf873b4e4c4c9a24879bceaa52dbe762d111fbd0604d1ff85c pre-commit
Time: 2021-03-24T16:47:16.702601
All clear. No secrets detected.
```

The problem that I believe is there is that tartufo does not see the real key included in the file, by error,
because there is an exclusion signature covering the first string and the newly added string is a concatenation
of previous one and some new random characters.

Of course if tartufo is run without signature exclusion it finds out both strings:

```
$ tartufo pre-commit
~~~~~~~~~~~~~~~~~~~~~
Reason: High Entropy
Filepath: howto.txt
Signature: c0c811fad6dddbaf873b4e4c4c9a24879bceaa52dbe762d111fbd0604d1ff85c
@@ -0,0 +1,7 @@
+This foobar service requires that all API keys start with this string:
+zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST
+
+
+You can use this key for example:
+zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST_xR6uDIpycfCfk3yUx
+

~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~
Reason: High Entropy
Filepath: howto.txt
Signature: c0c811fad6dddbaf873b4e4c4c9a24879bceaa52dbe762d111fbd0604d1ff85c
@@ -0,0 +1,7 @@
+This foobar service requires that all API keys start with this string:
+zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST
+
+
+You can use this key for example:
+zsNVsGF2/as4zJICuEIFDe0a0/wwMF2Go8eYcDST_xR6uDIpycfCfk3yUx
+

~~~~~~~~~~~~~~~~~~~~~
```

One may say that the human is the error here, it shouldn't have added the signature exclusion on the prefix.
But, even if this example is kind of convoluted, I think it can totally happen.

## Describe the solution you'd like

The signature is currently computed over the string and the filename, as in:

```
return blake2s("{}$${}".format(snippet, filename).encode("utf-8")).hexdigest()
```

which explains the above observed behavior.

Maybe the signature should take into account the position of the string in file,
either by line number or bytes offset from beginning. 
That way, if any single string appears multiple times in the file, it would need one exclusion signature
per instance, not per file. I guess this might also create problems for other use cases.
But the current computation is thus not more fine grained that the file based exclusions cases, that exclude
a whole file, the current signature exclusion excludes a given string in a given file, everywhere in the file.
I don't know if that is fully as designed, or just a side consequence.

Note also, that I am not myself convinced 100% that even with a line number/byte offset taken into account,
there are no possible other cases of the same behavior, so for now it is just a suggestion to enhance
the signature by using more than the file name.

## Describe alternatives you've considered

No alternatives, except human overview of the file content.

## Teachability, Documentation, Adoption, Migration Strategy

If the signature format is changed, for the same train of thoughts as in https://github.com/godaddy/tartufo/issues/176
there should be the time to include some kind of "signature version" to be able to run tartufo with various
versions of signatures if a new version is run with a configuration that was written with an old version and hence that may have signatures exclusions using a previous format.
Alternatively if the results of scans include the "tartufo version", the "signature version" could be infered from that, so there may be no need of a specific separate one.

**Comment by tarkatronic** [comment]
There was already [fairly involved design discussion](https://github.com/godaddy/tartufo/issues/2) on this feature, and the current design was decided upon after discussing several options, this one included.

TL;DR: As a file changes, so will the line number where a detected string exists, meaning every time the file is updated, the signature has to be updated as well, rendering them effectively useless.

**Comment by pmevzek-godaddy** [comment]
Thanks, I see the discussion, but I am not sure to understand why it converged immediately on string+filePath only where you presented yourself a first idea using line numbers.
I don't see any real explanations of why using the filename only is good enough. Except as you say that line number/byte offset can indeed change if the file is updated rendering previous signatures not matching anymore... but so it is also now if the filename is changed/moved. So I guess there is always a balance to choose between simplicity and security, but here simplicity was chosen over security then. Which is fine, as long as it is documented. Maybe some warnings on this situation should be written somewhere in the documentation to make it clear that all instances of a specific string  in the file will be covered by a single signature so one has to be cautious of secrets that can reuse parts of other secrets.
